### ISSUE TITLE
App env vars and service binding changes aren't reverted for failed rolling deployment

### ISSUE THREAD

**Issue by beyhan** [cloudfoundry/cloud_controller_ng#3531]
## Issue

A rolling update which  is [canceled](https://docs.cloudfoundry.org/devguide/deploy-apps/rolling-deploy.html#cancel) because of an application issue (e.g. failed start) will revert to the latest known droplet version but any changes to environment variables or service bindings will not be reverted. 
## Steps to Reproduce

1. Deploy a [slightly modified version](https://github.com/beyhan/spring-music/tree/add-version-and-env-var) of the [spring-music](https://github.com/cloudfoundry-samples/spring-music) app which introduces a `version` and a new env var called `TEST` by following the instructions available on the readme.
2. Verify that `https://<apphost>/appinfo` shows `{"profiles":["http2"],"services":[],"version":"1.0.0"}` and `https://<apphost>/envVarTest` shows `false`
3. Create a service instance from any type with the name `test`
4. Deploy with `cf push --strategy rolling` [another version](https://github.com/beyhan/spring-music/tree/broken-version-with-new-service) of the [spring-music](https://github.com/cloudfoundry-samples/spring-music) app which can't start and introduces changes to the app `version`, the value of the `TEST` env var and a binding to the `test` service instance.
5. The `cf push --strategy rolling` will fail and with `cf droplets <my-app>` you can check that the latest droplet isn't the current one, but `cf env <my-app>` will show the updated version of the env var and that the app is bound to the `test` service.
6. Restart the app with `cf restart <my-app>` Verify that `https://<apphost>/appinfo` shows `{"profiles":["http2"],"services":["test"],"version":"1.0.0"}` and `https://<apphost>/envVarTest` shows `true`.

## Expected result

The env var and service binding changes are not applied for failed rolling deployments.

## Current result

Currently the env var and service binding changes are applied to the application.

## Possible Fix

One option could be to use the [Cloud Foundry API app revisions](https://docs.cloudfoundry.org/devguide/revisions.html) to revert to the last known good version in such a situation.

**Comment by Rand0mF** [comment]
I agree with beyhan. For a normal cf push that's fine, as it will not have an old version of the app running on failure. A rolling push should be atomic. 

Giving an example:
Apps can be inspected with `cf app <app>` and seeing an app which is perfectly stable running, however, since some time ago, another user (or automation) failed to apply a new version, it is now not possible to restart the existing app anymore, reason unknown for the current user (or automation). The old version of the app is not designed to work with the new env values from the new app version.  

The configuration (env) should be bound to the version it is deployed with. It is counter-intuitive for an end-user to have a rolling deployment, which spins up a new version of the app but then applies the configuration to the old version of the app also. 

Rolling updates should be atomic, new configuration should only be applied for a new version, the old version should continue to run with the old configuration.
